xarray/plot/dataset_plot.py

```diff
@@ -0,0 +1,393 @@
+    label_from_attrs)
+
+# copied from seaborn
+_SCATTER_SIZE_RANGE = (np.r_[.5, 2]
+                       * np.square(mpl.rcParams["lines.markersize"]))
+
+
+def _infer_meta_data(ds, x, y, hue, hue_style, add_guide):
+    dvars = set(ds.variables.keys())
+    error_msg = (' must be one of ({0:s})'
+                 .format(', '.join(dvars)))
+
+    if x not in dvars:
+        raise ValueError('x' + error_msg)
+
+    if y not in dvars:
+        raise ValueError('y' + error_msg)
+
+    if hue is not None and hue not in dvars:
+        raise ValueError('hue' + error_msg)
+
+    if hue:
+        hue_is_numeric = _is_numeric(ds[hue].values)
+
+        if hue_style is None:
+            hue_style = 'continuous' if hue_is_numeric else 'discrete'
+
+        if not hue_is_numeric and (hue_style == 'continuous'):
+            raise ValueError('Cannot create a colorbar for a non numeric'
+                             ' coordinate: ' + hue)
+
+        if add_guide is None:
+            add_colorbar = True if hue_style == 'continuous' else False
+            add_legend = True if hue_style == 'discrete' else False
+
+    else:
+        if add_guide is True:
+            raise ValueError('Cannot set add_guide when hue is None.')
+        add_legend = False
+        add_colorbar = False
+
+    if hue_style is not None and hue_style not in ['discrete', 'continuous']:
+        raise ValueError('hue_style must be either None, \'discrete\' '
+                         'or \'continuous\'.')
+
+    if hue:
+        hue_label = label_from_attrs(ds[hue])
+        hue_values = ds[hue].values
+    else:
+        hue_label = None
+        hue_values = None
+
+    return {'add_colorbar': add_colorbar,
+            'add_legend': add_legend,
+            'hue_label': hue_label,
+            'hue_style': hue_style,
+            'xlabel': label_from_attrs(ds[x]),
+            'ylabel': label_from_attrs(ds[y]),
+            'hue_values': hue_values}
+
+
+def _infer_scatter_data(ds, x, y, hue, scatter_size, size_norm,
+                        size_mapping=None):
+
+    broadcast_keys = ['x', 'y']
+    to_broadcast = [ds[x], ds[y]]
+    if hue:
+        to_broadcast.append(ds[hue])
+        broadcast_keys.append('hue')
+    if scatter_size:
+        to_broadcast.append(ds[scatter_size])
+        broadcast_keys.append('size')
+
+    broadcasted = dict(zip(broadcast_keys, broadcast(*to_broadcast)))
+
+    data = {'x': broadcasted['x'],
+            'y': broadcasted['y'],
+            'hue': None,
+            'sizes': None}
+
+    if hue:
+        data['hue'] = broadcasted['hue']
+
+    if scatter_size:
+        size = broadcasted['size']
+
+        if size_mapping is None:
+            size_mapping = _parse_size(size, size_norm)
+
+        if _is_numeric(size):
+            # TODO : is there a better vectorized way of doing
+            # data['sizes'] = np.array([size_mapping.get(s0) for s0 in ss])
+            map_keys = np.array(list(size_mapping.keys()))
```

Comment:
need help here.

Reply:
You could switch `size_mappings` to use a pandas.Series, e.g., ` dict(zip(levels, widths))` -> `pandas.Series(widths, levels)`.

Then this becomes `data['sizes'] = size_mappings.loc[ss].values`.

Reply:
Thanks, that works well.